tests/kernel/semaphore/semaphore_error_cases/src/main.c

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (c) 2016, 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <irq_offload.h>
+
+#define STACK_SIZE (512 + CONFIG_TEST_EXTRA_STACKSIZE)
+#define THREAD_TEST_PRIORITY 0
+
+static struct k_thread tdata;
+static K_THREAD_STACK_DEFINE(tstack, STACK_SIZE);
+
+static struct k_sem sema;
+
+/* action after self-defined fatal handler. */
+static ZTEST_BMEM volatile bool valid_fault;
+static ZTEST_BMEM volatile bool assert_in_isr_has_done;
+
+
+static inline void set_fault_valid(bool valid)
+{
+	valid_fault = valid;
+}
+
+void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *pEsf)
+{
+	printk("Caught system error -- reason %d %d\n", reason, valid_fault);
+
+	if (valid_fault) {
+		printk("Fatal error expected as part of test case.\n");
+		valid_fault = false; /* reset back to normal */
+	} else {
+		printk("Fatal error was unexpected, aborting...\n");
+		k_fatal_halt(reason);
+	}
+}
+
+/* action for self-defined assert handler. */
+static ZTEST_DMEM volatile bool valid_assert;
+
+static inline void set_assert_valid(bool valid)
+{
+	valid_assert = valid;
+}
+
+static void action_after_assert_fail(void)
+{
+	/* The case which use irq_offload to enter interrupt context. */
+	if (k_is_in_isr()) {
+		k_sem_give(&sema);
+
+		/* Here we make assert fail trigger in ISR, but make it
+		 * continue to run util return from ISR.
+		 */
+		assert_in_isr_has_done = true;
+	}
+}
+
+#ifdef CONFIG_ASSERT_NO_FILE_INFO
+void assert_post_action(void)
+#else
+void assert_post_action(const char *file, unsigned int line)
+#endif
+{
+#ifndef CONFIG_ASSERT_NO_FILE_INFO
+	ARG_UNUSED(file);
+	ARG_UNUSED(line);
+#endif
+
+	printk("Caught assert failed\n");
```

Comment:
I guess that it should be "caught an assert.". Am I right?

Reply:
Yes! You are right